src/handle.rs

```diff
@@ -49,6 +49,12 @@ impl<Db: HasStorage> Handle<Db> {
         let storage = self.db.storage();
         storage.runtime().set_cancellation_flag();
 
+        self.db.salsa_event(Event {
```

Comment:
Nit: We could probably use an `AtomicUsize` for `clones`

Reply:
Yes, we could, we do need a mutex for signalling. This is simpler.